client/js/s3/request-signer.worker-manager.js

```diff
@@ -0,0 +1,88 @@
+/* globals qq, CryptoJS */
+
+/**
+ * Manages creation and communication of s3 signature workers.
+ */
+qq.s3.RequestSignerWorkerManager = function (o) {
+    "use strict";
+    var _worker = null,
+        _workerPromises = {},
+        options = {
+            workerUrl: null,
+            log: function(str, level) {}
+        };
+    qq.extend(options, o, true);
+
+    function init() {
+        var workerUrl;
+        switch (typeof options.workerUrl) {
+            case "string":
+                if (options.workerUrl !== "inline") {
+                    workerUrl = options.workerUrl;
+                } else {
+                    if (!qq.s3.createS3InlineWorkerUrl) {
+                        qq.Error("Missing inline s3 worker");
+                        return;
+                    }
+                    workerUrl = qq.s3.createS3InlineWorkerUrl();
+                }
+                break;
+            case "function":
+                workerUrl = options.workerUrl();
+                break;
+            default:
+                break;
+        }
+        if (!workerUrl) {
+            return;
+        }
+        try {
+            _worker = new Worker(workerUrl);
+            _worker.onerror = function (e) {
+                // Prevent the event from bubbling
+                e.preventDefault();
+                // log the error, and fail any pending promises.
+                options.log("Worker encountered an error. Disabling. " + e.message, "warn");
+                _worker = null;
+                var outstandingRequests = Object.keys(_workerPromises),
+                    i;
+                for (i = 0; i < outstandingRequests.length; i++) {
+                    _workerPromises[outstandingRequests[i]].failure(e);
+                    delete _workerPromises[outstandingRequests[i]];
+                }
+            };
+            _worker.onmessage = function (e) {
+                if (!_workerPromises[e.data.id]) {
+                    options.log("Worker returned a result for an request we dont know about.");
+                    return;
+                }
+                if (e.data.err) {
```

Comment:
Generally speaking, I'm in favor of Dijkstra's approach - single exit. I'm not married to this though. This is just a suggestion.